lexibank_kitchensemitic.py

```diff
@@ -1,187 +1,110 @@
-                    # we'll ignore it.
-                    
-                    # there are also forms that have no lexeme but do have a cognate set...
-                    # have to ignore these for now. Can't think of what to do with them
-                    # otherwise:
-                    #    Mɛhri Dog --- D
-                    #    Mɛhri Dry (adj.) --- F
-                    #    Hebrew Grass --- C/J
-                    #    Hebrew New --- A
-                    #    Mɛhri Seed --- C
-                    #    Hebrew Three --- A
-                    #    Mɛhri Warm --- H
-                    #    Hebrew Ye --- A
-
-                    for form, cog in zip(forms, cogs):
-                        assert len(form), "Empty form %s - %r - %r - %r" % (lid, gloss, form, cog)
-                        assert len(cog), "Empty cognate %s - %r - %r - %r" % (lid, gloss, form, cog)
-                        
-                        cogid = '%s-%s' % (csid, cog)
-                        for row in ds.add_lexemes(
-                                Language_ID=lid,
-                                Parameter_ID=csid,
-                                Value=form,
-                                Source=src
-                            ):
-                            if cog != '-':
-                                ds.add_cognate(lexeme=row, Cognateset_ID=cogid, Source=['Kitchen2009'])
+    language_class = KSLanguage
+
+    form_spec = FormSpec(
+        brackets={}, separators="/,", missing_data=("---",), strip_inside_brackets=False
+    )
+
+    def cmd_download(self, args):
+        self.raw_dir.xls2csv("Semitic.Wordlists.xls")
+        self.raw_dir.xls2csv("Semitic.Codings.Multistate.xlsx")
+
+    def cmd_makecldf(self, args):
+        lexeme_rows = self.raw_dir.read_csv("Semitic.Wordlists.ActualWordlists.csv")
+        lexeme_header = lexeme_rows.pop(0)
+
+        cognate_table = make_cognate_table(
+            self.raw_dir.read_csv("Semitic.Codings.Multistate.Sheet1.csv")
+        )
+
+        languages, concepts = {}, {}
+
+        for concept in self.conceptlist.concepts.values():
+            concepts[concept.english] = concept.id
+
+        args.writer.add_concepts(id_factory=lambda c: c.id)
+
+        for language in self.languages:
+            languages[language["Name"]] = (language["ID"], language["Sources"].split(";"))
+
+        args.writer.add_languages()
+
+        for row in pb(lexeme_rows, desc=f"Build CLDF for {self.id}"):
+            row = dict(zip(lexeme_header, row))
+            gloss = row.pop("Gloss")
+
+            for lang, lexeme in row.items():
+                lid, src = languages[lang]
+                cogs = cognate_table[lang][RELABLE_WORDS.get(gloss, gloss.upper())]
+
+                # Note that we cannot use the internal splitting algorithm, because of individual
+                # problems in the source data cells. Thus, we call form_spec explicitly.
```

Comment:
All lexibankers, see how the form-spec is used here explicitly, to make up for a complex input file with cognates and forms in different tables of the original data.

Reply:
I really like form_spec, it makes the code much much clearer.